File: server/src/routes/coopRoutes.ts
import { Env, UserRecord } from "../types";
import { createJsonResponse, createErrorResponse } from "../utils/cors";
import { extractBearerToken, generateSecureToken } from "../utils/googleAuth";

/**
 * Handles POST /coop/create
 */
export async function handleCreateCoop(request: Request, env: Env): Promise<Response> {
  try {
    const token = extractBearerToken(request);
    const user = await env.DB.prepare("SELECT id, username FROM users WHERE auth_token = ?").bind(token).first<UserRecord>();

    if (!user) {
      return createErrorResponse("Authentication required", 401);
    }

    const body: any = await request.json();
    let friendUserId: number | null = null;

    if (body.friend_username) {
      const targetUsername = String(body.friend_username).trim().toLowerCase();
      const friend = await env.DB.prepare("SELECT id FROM users WHERE lower(username) = ?")
        .bind(targetUsername)
        .first<UserRecord>();
      if (friend) {
        friendUserId = Number(friend.id);
      }
    }

    const sessionId = generateSecureToken("coop_");
    const currentTime = Date.now() / 1000;

    await env.DB.prepare(
      "INSERT INTO coop_sessions (session_id, host_user_id, friend_user_id, is_active, started_at) VALUES (?, ?, ?, 1, ?)"
    )
      .bind(sessionId, user.id, friendUserId, currentTime)
      .run();

    return createJsonResponse({ success: true, session_id: sessionId });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to create Co-Op room", 500);
  }
}

/**
 * Handles GET /coop/active
 */
export async function handleGetActiveCoop(request: Request, env: Env): Promise<Response> {
  try {
    const token = extractBearerToken(request);
    const user = await env.DB.prepare("SELECT id FROM users WHERE auth_token = ?").bind(token).first<UserRecord>();

    if (!user) {
      return createErrorResponse("Authentication required", 401);
    }

    const { results } = await env.DB.prepare(`
      SELECT cs.session_id, u.username as host_username, cs.started_at FROM coop_sessions cs
      JOIN users u ON cs.host_user_id = u.id
      WHERE cs.is_active = 1 AND cs.host_user_id != ?
        AND (cs.friend_user_id IS NULL OR cs.friend_user_id = ?)
    `)
      .bind(user.id, user.id)
      .all();

    return createJsonResponse({ success: true, rooms: results });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to fetch active Co-Op rooms", 500);
  }
}

/**
 * Handles POST /coop/join
 */
export async function handleJoinCoop(request: Request, env: Env): Promise<Response> {
  try {
    const token = extractBearerToken(request);
    const user = await env.DB.prepare("SELECT id FROM users WHERE auth_token = ?").bind(token).first<UserRecord>();

    if (!user) {
      return createErrorResponse("Authentication required", 401);
    }

    const body: any = await request.json();
    const sessionId = String(body.session_id || "").trim();

    const room = await env.DB.prepare("SELECT * FROM coop_sessions WHERE session_id = ? AND is_active = 1")
      .bind(sessionId)
      .first();

    if (!room) {
      return createErrorResponse("Active Co-Op session not found", 404);
    }

    if (!room.friend_user_id) {
      await env.DB.prepare("UPDATE coop_sessions SET friend_user_id = ? WHERE session_id = ?")
        .bind(user.id, sessionId)
        .run();
    }

    return createJsonResponse({ success: true, message: "Joined Co-Op focus room" });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to join Co-Op session", 500);
  }
}

/**
 * Handles POST /coop/end
 */
export async function handleEndCoop(request: Request, env: Env): Promise<Response> {
  try {
    const token = extractBearerToken(request);
    const user = await env.DB.prepare("SELECT id FROM users WHERE auth_token = ?").bind(token).first<UserRecord>();

    if (!user) {
      return createErrorResponse("Authentication required", 401);
    }

    const body: any = await request.json();
    const sessionId = String(body.session_id || "").trim();

    await env.DB.prepare("UPDATE coop_sessions SET is_active = 0 WHERE session_id = ? AND host_user_id = ?")
      .bind(sessionId, user.id)
      .run();

    return createJsonResponse({ success: true, message: "Co-Op session ended" });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to end Co-Op session", 500);
  }
}

File: server/src/routes/coop_routes.py
"""
Co-Op Focus Session Router for Prodo FastAPI Backend.

Handles cooperative multiplayer focus rooms, allowlisting partners, active multipliers,
and room lifecycle management using raw Request JSON parsing.
"""

import time
import secrets
from typing import Optional
from fastapi import APIRouter, Request, Header
from fastapi.responses import JSONResponse

from ..database import query_one, query_all, execute_db
from .auth_routes import extract_bearer_token, parse_json_body

router = APIRouter(prefix="/coop", tags=["Co-Op Rooms"])


@router.post("/create")
async def create_coop_room(request: Request, authorization: Optional[str] = Header(None)):
    """
    Creates a new Co-Op Focus Session room and returns a unique 6-character room code.
    """
    token = extract_bearer_token(authorization)
    if not token:
        return JSONResponse(status_code=401, content={"success": False, "error": "Unauthorized."})

    user = query_one("SELECT * FROM users WHERE auth_token = ?", (token,))
    if not user:
        return JSONResponse(status_code=401, content={"success": False, "error": "User session invalid."})

    session_id = secrets.token_hex(3).upper()

    execute_db(
        """
        INSERT INTO coop_sessions (session_id, host_user_id, is_active, started_at)
        VALUES (?, ?, 1, ?)
        """,
        (session_id, user["id"], time.time())
    )

    return {"success": True, "session_id": session_id, "message": "Co-Op session initialized successfully."}


@router.get("/active")
async def get_active_coop_rooms():
    """
    Returns list of active public/discoverable Co-Op focus rooms.
    """
    sql = """
    SELECT c.session_id, u.username AS host_username, c.started_at
    FROM coop_sessions c
    JOIN users u ON c.host_user_id = u.id
    WHERE c.is_active = 1
    ORDER BY c.started_at DESC
    LIMIT 20
    """
    rooms = query_all(sql)
    return {"success": True, "rooms": rooms}


@router.post("/join")
async def join_coop_room(request: Request, authorization: Optional[str] = Header(None)):
    """
    Joins an existing active Co-Op focus session using the session ID.
    """
    token = extract_bearer_token(authorization)
    if not token:
        return JSONResponse(status_code=401, content={"success": False, "error": "Unauthorized."})

    user = query_one("SELECT * FROM users WHERE auth_token = ?", (token,))
    if not user:
        return JSONResponse(status_code=401, content={"success": False, "error": "User session invalid."})

    body = await parse_json_body(request)
    session_id = str(body.get("session_id") or "").upper()

    if not session_id:
        return JSONResponse(status_code=400, content={"success": False, "error": "Session ID is required."})

    session = query_one("SELECT * FROM coop_sessions WHERE session_id = ? AND is_active = 1", (session_id,))
    if not session:
        return JSONResponse(status_code=404, content={"success": False, "error": "Co-Op room not found or inactive."})

    execute_db(
        "UPDATE coop_sessions SET friend_user_id = ? WHERE session_id = ?",
        (user["id"], session_id)
    )

    return {"success": True, "session_id": session_id, "message": "Joined Co-Op session!"}


@router.post("/end")
async def end_coop_room(request: Request, authorization: Optional[str] = Header(None)):
    """
    Terminates an active Co-Op focus session room.
    """
    body = await parse_json_body(request)
    session_id = str(body.get("session_id") or "").upper()
    if session_id:
        execute_db("UPDATE coop_sessions SET is_active = 0 WHERE session_id = ?", (session_id,))
    return {"success": True, "message": "Co-Op session terminated."}
